### DISCUSSION TITLE
prepareForValidation don't works with arrays with dot notation

### DISCUSSION THREAD

**Discussion by LocalHeroPro** [laravel/framework#32937] [General]
- Laravel Version: 7.12.0
- PHP Version:  7.3.18
- Database Driver & Version:

### Description:
I try prepare data for validation using using provided docs: https://laravel.com/docs/7.x/validation#prepare-input-for-validation.
But arrays with dot notation doesn't work as  expected.

### Steps To Reproduce:
1. Send form data:
```php
$formData = [
    'sender' => [
        'postal_code' => '26-500',
        //
    ]
];
```
2. Pass via controller
```php
<?php

namespace App\Http\Controllers;

use App\Http\Requests\StoreRequest;

class FormController extends Controller
{
    public function store(StoreRequest $request)
    {
        $validated = $request->validated();
        //
    }
}
```
3. Pass via validation
```php
<?php

namespace App\Http\Requests;

use Illuminate\Foundation\Http\FormRequest;
use Illuminate\Support\Str;

class StoreRequest extends FormRequest
{
    public function authorize(): bool
    {
        return true;
    }

    public function rules(): array
    {
        return [
            'sender.postal_code' => 'required|string|max:10',
            //
        ];
    }

    protected function prepareForValidation()
    {
        $this->merge(
            [
                'sender.postal_code' => (Str::of($this->sender['postal_code'])->replaceMatches('/[^a-zA-Z0-9]++/', ''))->__toString()
            ]
        );
    }
}
```
4. After that we get in controller
```php
$validated => [
    'sender' => [
        'postal_code' => '26-500',
        'sender.postal_code' => '26500',
        //
    ]
];
```
I want override $sender['postal_code'], against that I get new key with proper value.

**Comment by netpok** [comment]
IMO this should be in the issues section rather than here.

**Comment by tdias25** [comment]
Any solutions?

**Reply by thihakyaw** [reply]
Yes. You can check my comment for a solution :D

**Comment by morcen** [comment]
In case anyone is still blocked by this, I resolved this by using `$this->getInputSource()->replace()`.

For the OP's problem, this can be solved by doing it like this:

```
protected function prepareForValidation()
{
    $sender = $this->sender;
    $sender['postal_code'] = (Str::of($sender['postal_code'])->replaceMatches('/[^a-zA-Z0-9]++/', ''))->__toString();

    $this->getInputSource()->replace(['sender' => $sender]);
}
```

It's just a workaround, hopefully we can have a permanent solution to this soon.

**Reply by RomanNur** [reply]
Thanks for this, I hope this issue is gonna be fixed soon.

**Reply by thihakyaw** [reply]
I solved it. You can check under the comment I posted. :D

**Comment by thihakyaw** [comment]
Hi I got a solution for this if anyone wondering how to do alternative without sweating that much.
It is simple with only a little tweak. Example `prepareForValidation()` method in `Request` file.

```
    protected function prepareForValidation()
    {
        $data = $this->toArray();

        data_set($data, 'additional_data.library_card', '123456');
        data_set($data, 'additional_data.favorite_quote', '');
        data_set($data, 'additional_data.vip_member', true);

        $this->merge($data);
    }
```

`$data = $this->toArray()` get the form input from request in array form.

```
data_set($data, 'additional_data.library_card', '123456');
data_set($data, 'additional_data.favorite_quote', '');
data_set($data, 'additional_data.vip_member', true);
```

this set the data into array.

`$this->merge($data);` this merge the manipulated data to the form request.

Hopefully, you guys understand me. If there is some issues or security vulnerabilities, please point me out. As far as I know, this one is very appropriate solution I found.

**Reply by robertmylne** [reply]
This would work if it had a callback function to manipulate the data at that array interval, yours only lets you replace it with a static value.

**Comment by RomanNur** [comment]
Here is another solution.
Since [laravel's documentation](https://laravel.com/docs/8.x/validation#preparing-input-for-validation) using array's merging, you can do something like this.

```
protected function prepareForValidation(): void
{
        $this->merge([
            'user' => array_merge(
                $this->user,
                [
                    'name' => trim($this->input('user.name'),
                    'age' => (int) $this->input('user.age'),
                ]
            ),
            'company' => array_merge(
                $this->company,
                [
                    'name' => trim($this->input('company.name'),
                ]
            )
        ]);
}
```

You can do whatever you want with input's data, here is just an exapmle of changing data if the name has dots.

**Reply by jfrosorio** [reply]
I was having the same problem and this worked for me! Thanks for the help!

**Comment by 3bood-kr** [comment]
guys help the database columns are email, phone and the input names are admission_email, admission_phone whenever i submit the form they show as null in the database for some reason. I used to use this method to do this type of thing but it doesn't work anymore am i missing something??
```
public function prepareForValidation()
    {
        $this->merge([
            'email' => $this->admission_email,
            'phone' => $this->admission_phone,
        ]);
    }
```

**Reply by jfrosorio** [reply]
Can you please provide the rules you're using on your Form Request?
In the meanwhile, please check if admission_email and admission_phone are fillable properties.

**Reply by 3bood-kr** [reply]
> Can you please provide the rules you're using on your Form Request? In the meanwhile, please check if admission_email and admission_phone are fillable properties.

Thanks for replying. Here are the rules:
```
    public function rules(): array
    {
        return [
            'admission_email' => 'required|email|unique:admissions,email',
            'admission_phone' => 'required|numeric|digits_between:10,14',
            'address' => 'required|string|max:255',
            'student.*.student_name' => 'required|string|max:255',
            'student.*.admission_category_id' => 'required|numeric|exists:admission_categories,id',
            'student.*.class_room_id' => 'required|numeric|exists:class_rooms,id',
            'student.*.gender' => 'required|string|max:255|in:male,female',
            'student.*.old_grade' => 'required|numeric|min:0|max:100',
        ];
    }
```
I got it fixed by using this function:
```
    public function validated($key = null, $default = null)
    {
        $validated =  parent::validated($key, $default);
        return array_merge($validated, [
            'email' => $this->admission_email,
            'phone' => $this->admission_phone,
        ]);
    }
```
But i still think that prepareForValidation is cleaner but for some reason it doesn't work anymore for me even on the other projects i'm working on.
